### PR TITLE
[Merged by Bors] - implement genesis config

### LINKED ISSUE #3505: cmd/node: configuration changes for genesis id
part of: https://github.com/spacemeshos/pm/issues/135
depends on: https://github.com/spacemeshos/go-spacemesh/issues/3505

- add a genesis section into config
- for this issue genesis section needs to include: genesis time and genesis extra data (other fields can be moved into this config later)
   - genesis time is a string with time encoded according to RFC3339
   - genesis extra data is a string with a max size of 255 chars. defaults to `mainnet`
- when node is started check if genesis section is already stored in the data directory (example: /data/genesis.conf)
   - if it is not stored, encode genesis section with json and persist it on disk
   - if stored - decode object with json into another instance of genesis config, compute diff using https://github.com/google/go-cmp/ and print fatal error to the operator with a diff and a path to persisted genesis conf
- compute genesis id as a hash from genesis time and genesis extra data
- golden atx should be set to genesis id

### PR DESCRIPTION
**Pull request by dshulyak**
closes: https://github.com/spacemeshos/go-spacemesh/issues/3505

### CODE REVIEW

File: config/genesis.go [hunk 1]
```diff
@@ -1,15 +1,71 @@
...
+	hh.Write([]byte(g.ExtraData))
+	rst := types.Hash20{}
+	hh.Sum(rst[:0])
+	return rst
+}
+
+// Validate GenesisConfig.
+func (g *GenesisConfig) Validate() error {
+	if len(g.ExtraData) > 255 {
+		return fmt.Errorf("extra-data %s is longer than 255 symbols", g.ExtraData)
```
Comment (countvonzero) [inline config/genesis.go]: nit: probably better to print the extra-data at the end of the sentence.

File: cmd/node/node.go [hunk 2]
```diff
@@ -296,6 +299,31 @@ func (app *App) introduction() {
 
 // Initialize sets up an exit signal, logging and checks the clock, returns error if clock is not in sync.
 func (app *App) Initialize() (err error) {
+	// ensure all data folders exist
+	err = filesystem.ExistOrCreate(app.Config.DataDir())
```
Comment (countvonzero) [inline cmd/node/node.go:303]: for better usability maybe we should just go ahead to create the directory if it doesn't exist.

File: config/genesis.go [hunk 3]
```diff
@@ -1,15 +1,71 @@
...
+	return json.Unmarshal(buf, g)
+}
+
+// WriteToFile writes config content to file.
+func (g *GenesisConfig) WriteToFile(filename string) error {
+	buf, err := json.Marshal(g)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(filename, buf, 0600)
```
Comment (countvonzero) [inline config/genesis.go]: 0600 seems strict.... why not 0644?

File: cmd/node/node_test.go [hunk 4]
```diff
@@ -914,6 +914,90 @@ func TestConfig_GenesisAccounts(t *testing.T) {
...
+		app.Config = &config.Config{
+			BaseConfig: config.BaseConfig{
+				DataDirParent: t.TempDir(),
+			},
+			Genesis: &config.GenesisConfig{
+				ExtraData:   "test",
+				GenesisTime: time.Now().Format(time.RFC3339),
+			},
+		}
+		require.ErrorContains(t, app.Initialize(), "tortoise")
```
Comment (countvonzero) [inline cmd/node/node_test.go]: maybe you can start from DefaultTestConfig() and modify the genesis config and data dir so you don't hit the tortoise param error. see `TestInitialize_BadTortoiseParams`

File: cmd/node/node_test.go [hunk 5]
```diff
@@ -914,6 +914,90 @@ func TestConfig_GenesisAccounts(t *testing.T) {
...
+				DataDirParent: t.TempDir(),
+			},
+			Genesis: &config.GenesisConfig{
+				ExtraData:   "test",
+				GenesisTime: time.Now().Format(time.RFC3339),
+			},
+		}
+
+		require.ErrorContains(t, app.Initialize(), "tortoise")
+		require.ErrorContains(t, app.Initialize(), "tortoise")
```
Comment (countvonzero) [inline cmd/node/node_test.go]: double-whammy :P

**Review by countvonzero — APPROVED** [review]

File: cmd/node/node_test.go (hunk 4 above)
Comment (dshulyak) [reply cmd/node/node_test.go]: thanks, slipped from my mind

File: cmd/node/node.go (hunk 2 above)
Comment (dshulyak) [reply cmd/node/node.go:303]: but this function is creating a directory exactly in such case. did i miss your point?

Comment (countvonzero) [reply cmd/node/node.go:303]: lol. my bad. 

the error msg is "ensure folders exist" which misled me and prolly the operator. but i guess it'll print filesystem error to give the operator more hint.

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**